what4/src/What4/Utils/AbstractDomains.hs

```diff
@@ -585,14 +600,113 @@ natRangeToRange :: NatValueRange -> ValueRange Integer
 natRangeToRange (NatSingleRange x)  = SingleRange (toInteger x)
 natRangeToRange (NatMultiRange l u) = MultiRange (Inclusive (toInteger l)) (toInteger <$> u)
 
+------------------------------------------------------
+-- String abstract domain
+
+-- | The string abstract domain tracks an interval
+--   range for the length of the string.
+newtype StringAbstractValue =
+  StringAbs
+  { _stringAbsLength :: NatValueRange
+     -- ^ The length of the string falls in this range
+  }
+
+stringAbsTop :: StringAbstractValue
+stringAbsTop = StringAbs unboundedNatRange
+
+stringAbsEmpty :: StringAbstractValue
+stringAbsEmpty = StringAbs (natSingleRange 0)
+
+stringAbsJoin :: StringAbstractValue -> StringAbstractValue -> StringAbstractValue
+stringAbsJoin (StringAbs lenx) (StringAbs leny) =
+  StringAbs (natJoinRange lenx leny)
+
+stringAbsSingle :: StringLiteral si -> StringAbstractValue
+stringAbsSingle lit = StringAbs (natSingleRange (stringLitLength lit))
+
+stringAbsOverlap :: StringAbstractValue -> StringAbstractValue -> Bool
+stringAbsOverlap (StringAbs lenx) (StringAbs leny) = avOverlap BaseNatRepr lenx leny
+
+stringAbsCheckEq :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+stringAbsCheckEq (StringAbs lenx) (StringAbs leny)
+  | Just 0 <- asSingleNatRange lenx
+  , Just 0 <- asSingleNatRange leny
+  = Just True
+
+  | not (avOverlap BaseNatRepr lenx leny)
+  = Just False
+
+  | otherwise
+  = Nothing
+
+stringAbsConcat :: StringAbstractValue -> StringAbstractValue -> StringAbstractValue
+stringAbsConcat (StringAbs lenx) (StringAbs leny) =
+  StringAbs (natRangeAdd lenx leny)
+
+stringAbsSubstring :: StringAbstractValue -> NatValueRange -> NatValueRange -> StringAbstractValue
+stringAbsSubstring (StringAbs s) off len = StringAbs (natRange lo hi)
+ where
+ rng = addRange (natRangeToRange s) (negateRange (natRangeToRange off))
+
+ lo  = case rangeLowBound rng of
+         Unbounded -> 0
+         Inclusive x
+           | x < 0 -> 0
+           | otherwise -> min (fromInteger x) (natRangeLow len)
+
+ hi  = case (rangeHiBound rng, natRangeHigh len) of
+         (Unbounded  , y) -> y
+         (Inclusive x, Unbounded)   -> Inclusive (fromInteger (max 0 x))
+         (Inclusive x, Inclusive y) -> Inclusive (min (fromInteger (max 0 x)) y)
```

Comment:
Maybe this would be easier to understand if we defined a separate function `natRangeSub` that is an abstract version of subtraction on nats, with clamping at 0. Also another separate function `natRangeMin` that is an abstract version of `min` on nats. Then you could define `stringAbsSubstring` very simply as:
```
stringAbsSubstring (StringAbs s) off len = natRangeMin (natRangeSub s off) len
```